node/network/bridge/src/network.rs

```diff
@@ -66,11 +67,13 @@ pub(crate) fn send_message<M>(
 	// list. The message payload can be quite large. If the underlying
 	// network used `Bytes` this would not be necessary.
 	let last_peer = peers.pop();
+	// optimization: generate the protocol name once.
+	let protocol_name = protocol_names.get_name(peer_set, version);
```

Comment:
Generating this every time we send a message is really a bad idea :see_no_evil:

Reply:
Will fix

Reply:
Well, this time the `PeerSet` protocol versions are hardcoded throughout the codebase. This makes using a map `(protocol, version) -> protocol_name` a possible source of runtime errors. I'm thinking of a way to safeguard us from using any other versions of protocols than the ones registered in `PeerSetProtocolNames::new()`.

The best I can think of is creating the `ValidationVersion` & `CollationVersion` enums and making sure we have registered all the versions listed there. We can then use these enum variants in, e.g., `send_validation_message_v1()`. Nevertheless, these enum variants must eventually be converted into numbers, and I can't think of any other way than implementing a custom trait for version to number conversion. This must be a custom trait to make sure we only use our enums for version referencing. Nevertheless, this won't catch the errors of using `ValidationVersion` in place of `CollationVersion` and vice versa.

Reply:
To be honest, I don't like the solution proposed above. We'd have enums for version representation in a couple of places, but we'd still have numbers in protocol/PeerSet-agnostic parts. Do you see a better alternative @bkchr @eskimor ?

Reply:
Hmm this entire versioning story in the crate looks fishy :see_no_evil: If a peer is connected on version 1 and you would send him data on version 2, this wouldn't work :see_no_evil: 

> Nevertheless, these enum variants must eventually be converted into numbers, and I can't think of any other way than implementing a custom trait for version to number conversion

Why? To automate the generation of the protocol names? You could implement `From<ValidationVersion> for u32` and then call `validation_version.into()`. This all sounds a little bit "hacky", but as I said above the current situation of versioning is not great in the crate either.

Reply:
> You could implement From<ValidationVersion> for u32 and then call validation_version.into().

This won't allow us to support passing of both `ValidationVersion` and `CollationVersion` otherwise as `u32`, so an API user might end up passing numeric constants in places where we want to limit them only to supported versions. With a custom trait the user would be at least limited to passing `ValidationVersion` __or__ `CollationVersion`. (But still could pass one instead of the other.)

Reply:
Added protocol name lookup instead of generating it every time in https://github.com/paritytech/polkadot/pull/5876/commits/0ec120b363b8dc6d24f699ce33dc0038ffedd464